### PR TITLE
Fixed File.extname edge case for '.' in path with no extension

### PR DESCRIPTION
**Pull request by codyjb**
Fixes https://github.com/crystal-lang/crystal/issues/5780

### CODE REVIEW

**Comment by Sija** [comment]
Duplicate of #5782?

**Comment by codyjb** [comment]
Different implementation, also wasn't mentioned in the issue afaik. I'll leave it to the approver to decide which goes in 🙂

**Comment by straight-shoota** [comment]
I referenced the issue: https://github.com/crystal-lang/crystal/issues/5780#ref-pullrequest-302850315

But I think  your solution is preferable. It doesn't allocate a new string and is more performant.

**Comment by Sija** [comment]
This reads horribly though:

```cr
if dot_index && dot_index > separator_index && dot_index != filename.size - 1 && filename[dot_index - 1] != SEPARATOR
```

**Comment by straight-shoota** [comment]
Yes, that could be split up. It only needs to rindex the separator if the other conditions are true. This could optimize both code and performance.

**Comment by codyjb** [comment]
Absolutely, initial thought to implement, but it feels _not great_. Thoughts?

```ruby
  def self.extname(filename) : String
    filename.check_no_null_byte

    dot_index = filename.rindex('.')

    if dot_index && dot_index != filename.size - 1 && filename[dot_index - 1] != SEPARATOR
      if dot_index > (filename.rindex(SEPARATOR) || 0)
        filename[dot_index, filename.size - dot_index]
      else
        ""
    else
      ""
    end
  end
```

**Comment by Sija** [comment]
Shorter version (early returns FTW) but still not _that great_...

```cr
 def self.extname(filename) : String
    filename.check_no_null_byte

    dot_index = filename.rindex('.')

    if dot_index && dot_index != filename.size - 1 && filename[dot_index - 1] != SEPARATOR
      if dot_index > (filename.rindex(SEPARATOR) || 0)
        return filename[dot_index, filename.size - dot_index]
      end
    end

    ""
  end
```

**Comment by codyjb** [comment]
Its less bad than my last one, shall I update to match?

**Comment by straight-shoota** [comment]
You can actually get rid of `filename[dot_index - 1] != SEPARATOR` and replace it with `dot_index - 1 > (filename.rindex(SEPARATOR) || 0)`. It should come to the same result.

**Comment by straight-shoota** [comment]
While your at it, you should also add some specs for paths without separator to make sure they're properly handled.
```crystal
File.extname("test.cr").should eq(".cr")
File.extname("test.cr.cz").should eq(".cz")
File.extname(".test").should eq("")
File.extname(".test.cr").should eq(".cr")
File.extname(".test.cr.cz").should eq(".cz")
```

**Comment by codyjb** [comment]
Ah yes, thats much better! 

Will update with new test cases as well

File: src/file.cr [hunk 1]
```diff
@@ -297,13 +297,11 @@ class File < IO::FileDescriptor
...
 
     dot_index = filename.rindex('.')
-    separator_index = filename.rindex(SEPARATOR) || 0
 
-    if dot_index && dot_index > separator_index && dot_index != filename.size - 1 && filename[dot_index - 1] != SEPARATOR
-      filename[dot_index, filename.size - dot_index]
-    else
-      ""
+    if dot_index && dot_index != filename.size - 1 && dot_index - 1 > (filename.rindex(SEPARATOR) || 0)
+      return filename[dot_index, filename.size - dot_index]
```
Comment (bew) [inline src/file.cr]: Just remove the return here, and add an else for the empty string

**Comment by ysbaddaden** [comment]
Thank you!

**Pull request merged**